libraries/freertos_plus/standard/freertos_plus_cli/uart/FreeRTOS_CLI_UART.c

```diff
@@ -0,0 +1,65 @@
+/* FreeRTOS includes. */
+#include "FreeRTOS.h"
+#include "task.h"
+
+#include "FreeRTOS_CLI_UART.h"
+#include "iot_uart.h"
+
+extern IotUARTHandle_t xConsoleUart;
+#define cmdTIME_OUT_MS    ( 500 )
+
+int32_t uart_read( char * const pcInputBuffer,
+                   uint32_t xInputBufferLen );
+void uart_write( const char * const pcOutputBuffer,
+                 uint32_t xOutputBufferLen );
+
+xConsoleIO_t uartConsoleIO =
+{
+    .read  = uart_read,
+    .write = uart_write
+};
+
+int32_t uart_read( char * const pcInputBuffer,
+                   uint32_t xInputBufferLen )
+{
+    int32_t status;
+    uint16_t readSize = 0;
+
+    /* Read to input buffer. */
+    status = iot_uart_read_async( xConsoleUart, ( uint8_t * ) pcInputBuffer, xInputBufferLen );
+
+    if( status == IOT_UART_SUCCESS )
+    {
+        vTaskDelay( pdMS_TO_TICKS( cmdTIME_OUT_MS ) );
```

Comment:
Another potential solution is to register a callback and release a semaphore that is waiting here, instead of always waiting

Reply:
We can't use async APIs as it blocks until the complete length of input is received, which we can't predict as user can type in variable number of characters.
After some thoughts, I replaced the async call with a sync call reading one character at a time. This has two advantages:
1. The task can block until a character is pressed (which is most often the case).
2. The typed character is echoed back to console immediately thereby increasing responsiveness of CLI.